code/game/objects/items/trader.dm

```diff
@@ -752,3 +752,175 @@
+	name = "wet/dry vacuum"
+	desc = "A powerful vacuum cleaner that can collect both trash and fluids."
+	density = TRUE
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "wetdryvac1"
+	var/max_trash = 50
+	var/list/trash = list()
+	var/obj/item/vachandle/myhandle
+
+/obj/structure/wetdryvac/New()
+	..()
+	create_reagents(50)
+	myhandle = new /obj/item/vachandle(src)
+
+/obj/structure/wetdryvac/Destroy()
+	if(myhandle.loc == src)
+		qdel(myhandle)
+	else
+		myhandle.myvac = null
+	myhandle = null
+	for(var/obj/item/I in trash)
+		qdel(I)
+	trash.Cut()
+	..()
+
+/obj/structure/wetdryvac/examine(mob/user)
+	..()
+	to_chat(user,"<span class='info'>The wet tank gauge reads: [reagents.total_volume]/[reagents.maximum_volume]</span>")
+	to_chat(user,"<span class='info'>The dry storage gauge reads: [trash.len]/[max_trash]</span>")
+
+/obj/structure/wetdryvac/attackby(obj/item/W, mob/user)
+	if(istype(W,/obj/item/vachandle))
+		if(!myhandle)
+			myhandle = W
+		if(myhandle == W)
+			to_chat(user,"<span class='notice'>You insert \the [W] into \the [src].")
+			user.drop_item(W,src)
+			update_icon()
+	else
+		..()
+
+/obj/structure/wetdryvac/attack_hand(mob/user)
+	if(myhandle.loc == src)
+		user.put_in_hands(myhandle)
+		update_icon()
+	else
+		..()
+
+/obj/structure/wetdryvac/update_icon()
+	icon_state = "wetdryvac[myhandle.loc == src]"
+
+/obj/structure/wetdryvac/MouseDropFrom(var/obj/O, src_location, var/turf/over_location, src_control, over_control, params)
+	if(!can_use(usr,O))
+		return
+	if(istype(O,/obj/structure/sink))
+		if(!reagents.total_volume)
+			to_chat(usr,"<span class='warning'>\The [src] wet tank is already empty!</span>")
+			return
+		playsound(src, 'sound/effects/slosh.ogg', 25, 1)
+		reagents.clear_reagents()
+		to_chat(usr, "<span class='notice'>You flush \the [src] wet contents down \the [O].</span>")
+	else if(istype(O,/obj/item/weapon/reagent_containers) && O.is_open_container())
+		if(!reagents.total_volume)
+			to_chat(usr,"<span class='warning'>\The [src] wet tank is already empty!</span>")
+			return
+		playsound(src, 'sound/effects/slosh.ogg', 25, 1)
+		to_chat(usr, "<span class='notice'>You pour \the [src] wet contents into \the [O].</span>")
+		reagents.trans_to(O.reagents,reagents.total_volume)
+	else if(istype(O,/obj/machinery/disposal))
+		if(!contents.len)
+			to_chat(usr,"<span class='warning'>\The [src] dry storage is already empty!</span>")
+			return
+		playsound(src, 'sound/effects/freeze.ogg', 25, 1) //this sounds like trash moving to me
+		for(var/obj/item/I in trash)
+			I.forceMove(O)
+			trash -= I
+		to_chat(usr, "<span class='notice'>You dump \the [src] dry contents into \the [O].</span>")
+
+/obj/structure/wetdryvac/MouseDropTo(atom/O, mob/user as mob)
+	if(!can_use(user,O))
+		return
+	whrr(get_turf(O))
+
+/obj/structure/wetdryvac/proc/whrr(var/turf/T)
+	if(!T)
+		return
+	playsound(src, 'sound/effects/vacuum.ogg', 25, 1)
+	for(var/obj/effect/decal/cleanable/C in T)
+		var/datum/reagent/R = C.reagent
```

Comment:
The reagent in the cleanable seem to be strings, not types. You would need to create a reagent datum from the string of the decal, I think.
https://github.com/vgstation-coders/vgstation13/blob/f9a648cab932440204222263871df888c6e8e6b5/__DEFINES/reagents.dm#L186

Reply:
add_reagent wants a string, the real mistake I made here was having R be a reagent datum at all, it should just be a string